decode.go

```diff
@@ -70,7 +70,7 @@ func (de *decoder) message(buf []byte, sval reflect.Value) error {
 	// Decode all the fields
 	fields := ProtoFields(sval.Type())
 	fieldi := 0
-	for len(buf) > 0 {
+	for len(buf) > 0 && fieldi < len(fields) {
```

Comment:
Wait, what??   Protobuf semantics for decoding structs, in my understanding, is that the decoder *always* consumes the entire buffer.  It's explicitly allowed for there to be more than one value for a given field: if it's a single-value-only field, then the last value wins.  

If you take two separate encodings of a struct, and concatenate them together, you're supposed to get another valid encoding of a struct, in which all single-value fields defined in the second struct-encoding will override all single-value fields in the first.  (For multi-value fields, i.e., variable-sequence fields in non-packed representation, the contents are supposed to get concatenated I believe.)  

I think I remember reading this all in the protobuf spec some time ago, although I don't know if it might have changed since then.  Did you recheck?

And what exact problem is this proposed check intended to fix?